File: src/main/java/fi/quanfoxes/assembler/builders/Assign.java
package fi.quanfoxes.assembler.builders;

import fi.quanfoxes.assembler.*;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Variable;
import fi.quanfoxes.parser.nodes.OperatorNode;
import fi.quanfoxes.parser.nodes.VariableNode;

public class Assign {
    private static boolean isVariable(Node node) {
        return node instanceof VariableNode;
    }

    /**
     * Builds an assign operation into instructions
     * @param node Assign node
     * @return Assign instructions
     */
    public static Instructions build(Unit unit, OperatorNode node) {
        Instructions instructions = new Instructions();

        Instructions right = References.value(unit, node.getRight());
        Instructions left = References.direct(unit, node.getLeft());
        
        instructions.append(right, left);
        instructions.append(new Instruction("mov", left.getReference(), right.getReference(), left.getReference().getSize()));

        if (isVariable(node.getLeft())) {
            
            Variable variable = ((VariableNode)node.getLeft()).getVariable();
            instructions.setReference(Value.getVariable(right.getReference(), variable));
        }

        return instructions;
    }
}

File: Zigzag/Assembler/Builders/Assign.cs
public static class Assign
{
	/// <summary>
	/// Builds assign operation
	/// </summary>
	/// <param name="unit">Unit used to operate</param>
	/// <param name="node">Assign node</param>
	/// <returns>Instructions for assign operation</returns>
	public static Instructions Build(Unit unit, OperatorNode node)
	{
		var instructions = new Instructions();

		//var references = References.Get(unit, instructions, node.Left, node.Right, ReferenceType.DIRECT, ReferenceType.VALUE);
		References.Get(unit, instructions, node.Left, node.Right, ReferenceType.DIRECT, ReferenceType.VALUE, out Reference destination, out Reference source);
		
		Memory.Move(unit, instructions, source, destination);

		if (node.Left.GetNodeType() == NodeType.VARIABLE_NODE)
		{
			var variable = (node.Left as VariableNode).Variable;
			instructions.SetReference(Value.GetVariable(source, variable));
		}

		return instructions;
	}
}